### PR TITLE
refactor: move trailing_gaps bit into cursor

### PR DESCRIPTION
**Pull request by wolffcm**
Part of #6973, but I'm breaking this out into a separate PR for easier review.

This just moves the `trailing_gaps` bit from the `GapFiller` struct into `Cursor`, which simplifies the logic around how to resume gap-filling when starting a new output batch.

I also removed some assertions from tests in `algo.rs`. Updating them was painful, and I did not find that verifying cursor state at intermediate steps was all that useful. There is still lots of coverage for all of these cases in `exec_tests.rs` in the `gapfill` module.

### CODE REVIEW

File: iox_query/src/exec/gapfill/algo.rs [hunk 1]
```diff
@@ -467,18 +447,25 @@ impl Cursor {
...
-
-        // Process the first series separately as it may just be a part of what
-        // did not fit in the previous output batch.
-        self.append_series_items(params, input_time_array, *first_series, &mut f)?;
+        for series in series_ends.iter() {
+            if self
+                .next_ts
+                .map_or(false, |next_ts| next_ts > params.last_ts)
+            {
+                self.next_ts = params.first_ts;
```
Comment (wolffcm) [inline iox_query/src/exec/gapfill/algo.rs:455]: The main reason for this refactoring is that now the "switch to the next series" logic all happens in a single place, which is right here. This makes it much simpler to implement `FILL(PREVIOUS)`.

File: iox_query/src/exec/gapfill/algo.rs [hunk 2]
```diff
@@ -317,17 +271,43 @@ struct Cursor {
...
+    /// Produces `None` for the case where `next_input_offset` is equal to `series_end`,
+    /// and there are no trailing gaps to produce.
     fn count_series_rows(
         &mut self,
         params: &GapFillParams,
         input_time_array: &TimestampNanosecondArray,
         series_end: usize,
-    ) -> usize {
+    ) -> Option<usize> {
+        if !self.trailing_gaps && self.next_input_offset == series_end {
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo.rs:307]: is this a change in behavior? Or just a refactoring of existing logic to here?

File: iox_query/src/exec/gapfill/algo.rs [hunk 3]
```diff
@@ -633,14 +616,7 @@ mod tests {
...
 
-        assert_eq!(
-            Cursor {
-                next_input_offset: input_times.len(),
-                next_ts: Some(params.last_ts + params.stride),
-                remaining_output_batch_size: output_batch_size - 7
-            },
-            cursor
-        );
+        assert_cursor_end_state(&cursor, &input_times, &params);
```
Comment (alamb) [inline iox_query/src/exec/gapfill/algo.rs:619]: I agree checking the intermediate cursor state seems redundant.

**Review by alamb — APPROVED** [review]
Makes sense to me 👍

File: iox_query/src/exec/gapfill/algo.rs (hunk 2 above)
Comment (wolffcm) [reply iox_query/src/exec/gapfill/algo.rs:307]: No change, the logic just moved from `plan_output_batch` to here.

**Pull request merged**